types/firestore.d.ts

```diff
@@ -1067,6 +1107,91 @@ declare namespace FirebaseFirestore {
      */
     isEqual(other: FieldPath): boolean;
   }
+
+  /**
+   * A Timestamp represents a point in time independent of any time zone or
+   * calendar, represented as seconds and fractions of seconds at nanosecond
+   * resolution in UTC Epoch time. It is encoded using the Proleptic Gregorian
+   * Calendar which extends the Gregorian calendar backwards to year one. It is
+   * encoded assuming all minutes are 60 seconds long, i.e. leap seconds are
+   * "smeared" so that no leap second table is needed for interpretation. Range
+   * is from 0001-01-01T00:00:00Z to 9999-12-31T23:59:59.999999999Z.
+   *
+   * @see https://github.com/google/protobuf/blob/master/src/google/protobuf/timestamp.proto
+   */
+  export class Timestamp {
+    /**
+     * Creates a new timestamp with the current date, with millisecond precision.
+     *
+     * @return A new `Timestamp` representing the current date.
+     */
+    static now(): Timestamp;
+
+    /**
+     * Creates a new timestamp from the given date.
+     *
+     * @param date The date to initialize the `Timestamp` from.
+     * @return A new `Timestamp` representing the same point in time as the
+     * given date.
+     */
+    static fromDate(date: Date): Timestamp;
+
+    /**
+     * Creates a new timestamp from the given number of milliseconds.
+     *
+     * @param milliseconds Number of milliseconds since Unix epoch
+     * 1970-01-01T00:00:00Z.
+     * @return A new `Timestamp` representing the same point in time as the
+     * given number of milliseconds.
+     */
+    static fromMillis(milliseconds: number): Timestamp;
+
+    /**
+     * Creates a new timestamp.
+     *
+     * @param seconds The number of seconds of UTC time since Unix epoch
+     * 1970-01-01T00:00:00Z. Must be from 0001-01-01T00:00:00Z to
+     * 9999-12-31T23:59:59Z inclusive.
+     * @param nanoseconds The non-negative fractions of a second at nanosecond
+     * resolution. Negative second values with fractions must still have
+     * non-negative nanoseconds values that count forward in time. Must be from
+     * 0 to 999,999,999 inclusive.
+     */
+    constructor(seconds: number, nanoseconds: number);
+
+    /**
+     * The number of seconds of UTC time since Unix epoch 1970-01-01T00:00:00Z.
+     */
+    readonly seconds: number;
```

Comment:
Hi,
Sorry for commenting on a merged pull request, but it seems like the js object has a field called "_seconds" and another "_nanoseconds".
Whenever I create a Timestamp by using Timestamp.now() in TS, it is creating an object with {seconds:..., nanoseconds: ...}, not having the underscores. When I try to save this to a firestore repository it fails with the following message:
"Error: Argument "data" is not a valid Document. Detected an object of type "Timestamp" that doesn't match the expected instance. Please ensure that the Firestore types you are using are from the same NPM package."
Should the ts definition have underscores in the names of these fields as well?

Reply:
Hi! Are you using the Admin SDK? If you are using the Admin SDK, you have to use the Admin-exported type of Timestamp (`admin.firestore.Timestamp`). That, however, is not possible until the next release of the Admin SDK, which will include the definition for this type (as per https://github.com/firebase/firebase-admin-node/pull/309).